cache/cache_test.go

```diff
@@ -0,0 +1,46 @@
+package cache_test
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/treeverse/lakefs/cache"
+)
+
+func TestCacheRace(t *testing.T) {
+	const (
+		parallelism = 25
+		n           = 200
+		worldSize   = 10
+		cacheSize   = 7
+	)
+
+	c := cache.NewCache(cacheSize, time.Hour*12, cache.NewJitterFn(time.Millisecond))
+
+	start := make(chan struct{})
+	wg := sync.WaitGroup{}
+
+	for i := 0; i < parallelism; i++ {
+		wg.Add(1)
+		go func(i int) {
+			<-start
+			for j := 0; j < n; j++ {
+				k := j % worldSize
+				kk, err := c.GetOrSet(k, func() (interface{}, error) {
+					return k * k, nil
+				})
+				if err != nil {
+					t.Error(err)
+					return
+				}
+				if kk.(int) != k*k {
+					t.Errorf("[%d] got %d^2=%d, expected %d", i, k, kk, k*k)
```

Comment:
goroutine can't call Error or Fatal

Reply:
So I thought so too, then I forgot, so now I had to check.  https://golang.org/pkg/testing/#T says I am allowed to call Error -- but _not_ Fatal -- in a goroutine:

>  A test ends when its Test function returns or calls any of the methods FailNow, Fatal, Fatalf, SkipNow, Skip, or Skipf. Those methods, as well as the Parallel method, must be called only from the goroutine running the Test function.
>
> The other reporting methods, such as the variations of Log and Error, may be called simultaneously from multiple goroutines. 

So "citation required"?

Reply:
You got it right, my bad - the call to Error will work